### PR TITLE
Improved permissions in the chat module.

### PR DESCRIPTION
**Pull request by arek1-618**
Improved permissions in the chat module.
Improved unit testing.
**Adding a link to the record.**

**Change in appearance.**
chrome_2018-11-14_13-59-03

**Disable chat if user is switched.**

### CODE REVIEW

File: app/Chat.php [hunk 1]
```diff
@@ -196,20 +196,27 @@ public static function getRoomsGroup(?int $userId = null)
 		if (empty($userId)) {
 			$userId = User::getCurrentUserId();
 		}
+		$groups = \App\Fields\Owner::getInstance('CustomView')->getGroups(false);
```
Comment (mariuszkrzaczkowski) [inline app/Chat.php]: add it as a new function

File: modules/Chat/views/Entries.php [hunk 2]
```diff
@@ -37,6 +44,24 @@ public function checkPermission(\App\Request $request)
 		if (!$currentUserPriviligesModel->hasModulePermission($request->getModule())) {
 			throw new \App\Exceptions\NoPermitted('ERR_NOT_ACCESSIBLE', 406);
 		}
+		if ($request->has('roomType') && !$request->has('recordId')) {
+			throw new \App\Exceptions\NoPermittedToRecord('ERR_NOT_ALLOWED_VALUE', 406);
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/views/Entries.php]: throw new \App\Exceptions\NoPermittedToRecord('ERR_NO_PERMISSIONS_FOR_THE_RECORD', 406);

File: public_html/layouts/basic/modules/Chat/resources/Chat.js [hunk 3]
```diff
@@ -702,11 +702,15 @@ window.Chat_JS = class Chat_Js {
...
 	 */
-	createRoomItem(data = {}, favorite = false) {
+	createRoomItem(data = {}, favorite = false, roomType = 'global') {
 		let itemRoom = this.container.find('.js-room-list .js-temp-item-room').clone(false, false);
 		itemRoom.removeClass('js-temp-item-room').removeClass('hide');
+		itemRoom.addClass('d-flex');
 		itemRoom.find('.js-room-name').html(data.name);
 		itemRoom.attr('title', data.name);
+		if (roomType === 'crm') {
+			itemRoom.find('.js-link').removeClass('hide').attr('href', "index.php?module=Contacts&view=Detail&record=" + data.recordid);
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Chat/resources/Chat.js]: Contacts ?

File: app/Chat.php [hunk 4]
```diff
@@ -88,14 +88,20 @@ public static function setCurrentRoom(string $roomType, ?int $recordId)
...
-			if ($row === false) {
-				return false;
-			}
-			return [
-				'roomType' => 'global',
-				'recordId' => $row[static::COLUMN_NAME['room']['global']]
-			];
+			return static::getDefaultRoom();
+		}
+		$recordId = $_SESSION['chat']['recordId'];
```
Comment (rskrzypczak) [inline app/Chat.php]: Why you don't use \App\Session?

File: app/Chat.php [hunk 5]
```diff
@@ -88,14 +88,20 @@ public static function setCurrentRoom(string $roomType, ?int $recordId)
...
+		switch ($_SESSION['chat']['roomType']) {
+			case 'crm':
+				if (!\Vtiger_Record_Model::getInstanceById($recordId)->isViewable()) {
```
Comment (rskrzypczak) [inline app/Chat.php]: What happens when the record does not exist.
You will get an exception.

File: app/Chat.php [hunk 6]
```diff
@@ -196,20 +202,27 @@ public static function getRoomsGroup(?int $userId = null)
 		if (empty($userId)) {
 			$userId = User::getCurrentUserId();
 		}
+		$groups = Fields\Owner::getUserGroups();
```
Comment (mariuszkrzaczkowski) [inline app/Chat.php]: ?

File: app/Chat.php [hunk 7]
```diff
@@ -733,14 +776,13 @@ public function getParticipants()
 		$dataReader = $query->createCommand()->query();
 		$participants = [];
 		while ($row = $dataReader->read()) {
-			$userModel = User::getUserModel($row['userid']);
-			$image = $userModel->getImage();
+			$user = $this->getUserInfo($row['userid']);
```
Comment (mariuszkrzaczkowski) [inline app/Chat.php]: User::getUserModel  ??$this->getUserInfo

File: app/Fields/Owner.php [hunk 8]
```diff
@@ -46,6 +46,16 @@ public static function getInstance($moduleName = false, $currentUser = false)
 		return $instance;
 	}
 
+	/**
+	 * Get user groups.
+	 *
+	 * @return string[]
+	 */
+	public static function getUserGroups()
```
Comment (mariuszkrzaczkowski) [inline app/Fields/Owner.php]: ?

File: modules/Accounts/models/DetailView.php [hunk 9]
```diff
@@ -122,7 +122,7 @@ public function getDetailViewRelatedLinks()
 				'linkicon' => 'fa-twitter',
 			];
 		}
-		if (\App\Module::isModuleActive('Chat') && $recordModel->isViewable()) {
+		if (\App\User::getCurrentUserId() === \App\User::getCurrentUserRealId() && \App\Module::isModuleActive('Chat') && $recordModel->isViewable()) {
```
Comment (mariuszkrzaczkowski) [inline modules/Accounts/models/DetailView.php]: unnecessary $recordModel->isViewable()

File: modules/Chat/views/Entries.php [hunk 10]
```diff
@@ -37,6 +44,27 @@ public function checkPermission(\App\Request $request)
 		if (!$currentUserPriviligesModel->hasModulePermission($request->getModule())) {
 			throw new \App\Exceptions\NoPermitted('ERR_NOT_ACCESSIBLE', 406);
 		}
+		if (\App\User::getCurrentUserId() !== \App\User::getCurrentUserRealId()) {
+			throw new \App\Exceptions\NoPermitted('ERR_NOT_ACCESSIBLE', 406);
+		}
+		if ($request->has('roomType') && !$request->has('recordId')) {
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/views/Entries.php]:
```suggestion
		if ($request->has('roomType') ||!$request->has('recordId')) {
```

File: layouts/basic/modules/Chat/Modal.tpl [hunk 11]
```diff
@@ -3,18 +3,25 @@
...
+			<div class="col-7 p-0">
 				<span class="js-room-name" data-js="append|replace">{$ROOM['name']}</span>
 			</div>
 			<div class="col-3 p-0 text-right">
 				<span class="js-room-cnt badge badge-info ml-1 inline" data-js="append|replace">
 					{if $ROOM['cnt_new_message'] > 0}{$ROOM['cnt_new_message']}{/if}
 				</span>
 			</div>
+			<div class="col-1 text-right px-2 d-flex align-items-center o-chat__pin-favorites">
+				<a href="index.php?module={if isset($ROOM['moduleName'])}{$ROOM['moduleName']}{/if}&view=Detail&record={$ROOM['recordid']}"
```
Comment (rskrzypczak) [inline layouts/basic/modules/Chat/Modal.tpl]: $ROOM['moduleName'] can't be empty.
check $ROOM_TYPE!=='crm' in the begining

File: tests/Base/Z_Chat.php [hunk 12]
```diff
@@ -276,12 +311,136 @@ public function testRemoveRecordCrm()
...
+		$this->assertNotFalse($keyUser, 'Problem with the method "getParticipants"');
+		$this->assertSame($participants[$keyUser]['message'], $entriesAfter[$key]['messages']);
+	}
+
+	/**
+	 * Testing retrieving information about rooms.
+	 *
+	 * @throws \App\Exceptions\AppException
+	 */
+	public function testRoomGetAll()
```
Comment (mariuszkrzaczkowski) [inline tests/Base/Z_Chat.php]: transfer to the selenium

File: tests/Base/Z_Chat.php [hunk 13]
```diff
@@ -276,12 +311,136 @@ public function testRemoveRecordCrm()
...
+	}
+
+	/**
+	 * Testing the sending method.
+	 *
+	 * @throws \App\Exceptions\AppException
+	 * @throws \App\Exceptions\IllegalValue
+	 * @throws \yii\db\Exception
+	 */
+	public function testSend()
```
Comment (mariuszkrzaczkowski) [inline tests/Base/Z_Chat.php]: transfer to the selenium

**Pull request merged**